In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
path = Path("Resources/lending_data.csv")
df_lending_data = pd.read_csv(path)
# Review the DataFrame
df_lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending_data['loan_status']
# Separate the X variable, the features
X = df_lending_data.drop('loan_status', axis=1)

In [5]:
# Review the y variable Series
y.shape

(77536,)

In [6]:
# Review the X variable DataFrame
X.shape

(77536, 7)

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [7]:
# Check the balance of our target values
balance = y.value_counts()
balance

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)
# Fit the model using training data
model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
y_pred = model.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


In [12]:
# Print the balanced_accuracy score of the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy Score: {accuracy:.2f}')


Accuracy Score: 0.99


In [13]:
# Generate a confusion matrix for the model
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

Confusion Matrix:
[[14924    77]
 [   31   476]]


In [14]:
# Print the classification report for the model
class_report = classification_report(y_test,y_pred)
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.86      0.94      0.90       507

    accuracy                           0.99     15508
   macro avg       0.93      0.97      0.95     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The model predicts values with a high level of accuracy. However, this should be tempered with the fact that the dataset is imbalanced. Running the balance test provided us with the following results: 0 75036;1 2500. Approximately 97% of the data is for healthy loans which could lead to a misleading evaluation and fail to detect the minority class

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [41]:
!pip install --upgrade scikit-learn
!pip install --upgrade imbalanced-learn



  Obtaining dependency information for scikit-learn from https://files.pythonhosted.org/packages/a8/e9/3e4879974a7c4dcaca2a746dde3df08d0ae8f14c74b03591616ce5f0a8b1/scikit_learn-1.4.0-1-cp311-cp311-win_amd64.whl.metadata
   ---------------------------------------- 0.0/10.6 MB ? eta -:--:--
   ---------------------------------------- 0.1/10.6 MB 1.7 MB/s eta 0:00:07
    --------------------------------------- 0.2/10.6 MB 2.0 MB/s eta 0:00:06
   - -------------------------------------- 0.3/10.6 MB 1.9 MB/s eta 0:00:06
   - -------------------------------------- 0.4/10.6 MB 1.9 MB/s eta 0:00:06
   - -------------------------------------- 0.5/10.6 MB 2.1 MB/s eta 0:00:05
   -- ------------------------------------- 0.6/10.6 MB 2.1 MB/s eta 0:00:05
   -- ------------------------------------- 0.7/10.6 MB 2.0 MB/s eta 0:00:05
   -- ------------------------------------- 0.8/10.6 MB 2.1 MB/s eta 0:00:05
   --- ------------------------------------ 0.9/10.6 MB 2.1 MB/s eta 0:00:05
   --- ----------

ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'C:\\Users\\Admin\\anaconda3\\Lib\\site-packages\\~klearn\\cluster\\_dbscan_inner.cp311-win_amd64.pyd'
Consider using the `--user` option or check the permissions.



  Obtaining dependency information for imbalanced-learn from https://files.pythonhosted.org/packages/a3/9e/fbe60a768502af54563dcb59ca7856f5a8833b3ad5ada658922e1ab09b7f/imbalanced_learn-0.11.0-py3-none-any.whl.metadata
   ---------------------------------------- 0.0/235.6 kB ? eta -:--:--
   ---------------------------------------- 0.0/235.6 kB ? eta -:--:--
   ---------- ----------------------------- 61.4/235.6 kB ? eta -:--:--
   ------------------- -------------------- 112.6/235.6 kB 1.7 MB/s eta 0:00:01
   ------------------------ --------------- 143.4/235.6 kB 1.2 MB/s eta 0:00:01
   ----------------------------- ---------- 174.1/235.6 kB 1.1 MB/s eta 0:00:01
   ---------------------------------------  235.5/235.6 kB 1.1 MB/s eta 0:00:01
   ---------------------------------------- 235.6/235.6 kB 1.0 MB/s eta 0:00:00
  Attempting uninstall: imbalanced-learn
    Found existing installation: imbalanced-learn 0.10.1
    Uninstalling imbalanced-learn-0.10.1:
      Successfully uninstall

In [15]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

In [18]:
# Count the distinct values of the resampled labels data
unique_values = y_resampled.value_counts()
unique_values

0    60035
1    60035
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [19]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logreg_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
logreg_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred = logreg_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [20]:
# Print the balanced accuracy score
balanced_acc_score = balanced_accuracy_score(y_test, y_pred)
print(f'Balanced Accuracy Score: {balanced_acc_score:.4f}')

# Generate a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Print the classification report
class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Balanced Accuracy Score: 0.9941
Confusion Matrix:
[[14914    87]
 [    3   504]]
Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.85      0.99      0.92       507

    accuracy                           0.99     15508
   macro avg       0.93      0.99      0.96     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model, trained with oversampled data, demonstrates excellent predictive performance for both healthy loans and high-risk loans. It is well-suited for the credit risk classification task, achieving high accuracy and effectively identifying instances of interest in the imbalanced dataset.